File: data/leetcode_solutions/solutions/0437. Path Sum III/0437.cpp
class Solution {
 public:
  int pathSum(TreeNode* root, int targetSum) {
    if (root == nullptr)
      return 0;
    return dfs(root, targetSum) +
           pathSum(root->left, targetSum) +
           pathSum(root->right, targetSum);
  }

 private:
  int dfs(TreeNode* root, int targetSum) {
    if (root == nullptr)
      return 0;
    return (targetSum == root->val) +
           dfs(root->left, targetSum - root->val) +
           dfs(root->right, targetSum - root->val);
  }
};

File: data/leetcode_solutions/solutions/0437. Path Sum III/0437.java
class Solution {
  public int pathSum(TreeNode root, int targetSum) {
    if (root == null)
      return 0;
    return dfs(root, targetSum) + pathSum(root.left, targetSum) + pathSum(root.right, targetSum);
  }

  private int dfs(TreeNode root, int targetSum) {
    if (root == null)
      return 0;
    return (targetSum == root.val ? 1 : 0) +
        dfs(root.left, targetSum - root.val) +
        dfs(root.right, targetSum - root.val);
  }
}
